// crisp-sdk-web/dist/crisp.umd.cjs
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
    typeof define === 'function' && define.amd ? define(['exports'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global["crisp-sdk-web"] = {}));
})(this, (function (exports) { 'use strict';

    /*
     * This file is part of crisp-sdk-web
     *
     * Copyright (c) 2025 Crisp IM SAS
     * All rights belong to Crisp IM SAS
     */
    /**************************************************************************
     * CLASS
     ***************************************************************************/
    /**
     * Crisp message management
     */
    var CrispMessage = /** @class */ (function () {
        /**
         * Constructor
         */
        function CrispMessage(crisp) {
            this.parent = crisp;
        }
        /**
         * Sets the message input text
         */
        CrispMessage.prototype.setMessageText = function (content) {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["set", "message:text", [content]]);
        };
        /**
         * Sends a message
         */
        CrispMessage.prototype.send = function (type, content) {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["do", "message:send", [type, content]]);
        };
        /**
         * Sends a text message
         */
        CrispMessage.prototype.sendText = function (content) {
            this.send("text", content);
        };
        /**
         * Sends a file message
         */
        CrispMessage.prototype.sendFile = function (content) {
            this.send("file", content);
        };
        /**
         * Sends an animation message
         */
        CrispMessage.prototype.sendAnimation = function (content) {
            this.send("animation", content);
        };
        /**
         * Sends an audio message
         */
        CrispMessage.prototype.sendAudio = function (content) {
            this.send("audio", content);
        };
        /**
         * Shows a message in the chat
         */
        CrispMessage.prototype.show = function (type, content) {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["do", "message:show", [type, content]]);
        };
        /**
         * Shows a text message
         */
        CrispMessage.prototype.showText = function (content) {
            this.show("text", content);
        };
        /**
         * Shows a file message
         */
        CrispMessage.prototype.showFile = function (content) {
            this.show("file", content);
        };
        /**
         * Shows an animation message
         */
        CrispMessage.prototype.showAnimation = function (content) {
            this.show("animation", content);
        };
        /**
         * Shows an audio message
         */
        CrispMessage.prototype.showAudio = function (content) {
            this.show("audio", content);
        };
        /**
         * Shows a picker message
         */
        CrispMessage.prototype.showPicker = function (content) {
            this.show("picker", content);
        };
        /**
         * Shows a field message
         */
        CrispMessage.prototype.showField = function (content) {
            this.show("field", content);
        };
        /**
         * Shows a carousel message
         */
        CrispMessage.prototype.showCarousel = function (content) {
            this.show("carousel", content);
        };
        /**
         * Marks all messages as read
         */
        CrispMessage.prototype.markAsRead = function () {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["do", "message:read"]);
        };
        /**
         * Starts a message thread
         */
        CrispMessage.prototype.startThread = function (name) {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["do", "message:thread:start", [name]]);
        };
        /**
         * Ends a message thread
         */
        CrispMessage.prototype.endThread = function (name) {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["do", "message:thread:end", [name]]);
        };
        /**
         * Registers a callback for message sent event
         */
        CrispMessage.prototype.onMessageSent = function (callback) {
            this.parent.createSingletonIfNecessary();
            this.offMessageSent();
            window.$crisp.push(["on", "message:sent", callback]);
        };
        /**
         * Unregisters the message sent callback
         */
        CrispMessage.prototype.offMessageSent = function () {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["off", "message:sent"]);
        };
        /**
         * Registers a callback for message received event
         */
        CrispMessage.prototype.onMessageReceived = function (callback) {
            this.parent.createSingletonIfNecessary();
            this.offMessageReceived();
            window.$crisp.push(["on", "message:received", callback]);
        };
        /**
         * Unregisters the message received callback
         */
        CrispMessage.prototype.offMessageReceived = function () {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["off", "message:received"]);
        };
        /**
         * Registers a callback for message compose sent event
         */
        CrispMessage.prototype.onMessageComposeSent = function (callback) {
            this.parent.createSingletonIfNecessary();
            this.offMessageComposeSent();
            window.$crisp.push(["on", "message:compose:sent", callback]);
        };
        /**
         * Unregisters the message compose sent callback
         */
        CrispMessage.prototype.offMessageComposeSent = function () {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["off", "message:compose:sent"]);
        };
        /**
         * Registers a callback for message compose received event
         */
        CrispMessage.prototype.onMessageComposeReceived = function (callback) {
            this.parent.createSingletonIfNecessary();
            this.offMessageComposeReceived();
            window.$crisp.push(["on", "message:compose:received", callback]);
        };
        /**
         * Unregisters the message compose received callback
         */
        CrispMessage.prototype.offMessageComposeReceived = function () {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["off", "message:compose:received"]);
        };
        return CrispMessage;
    }());

    /*
     * This file is part of crisp-sdk-web
     *
     * Copyright (c) 2025 Crisp IM SAS
     * All rights belong to Crisp IM SAS
     */
    /**************************************************************************
     * CLASS
     ***************************************************************************/
    /**
     * Crisp user management
     */
    var CrispUser = /** @class */ (function () {
        /**
         * Constructor
         */
        function CrispUser(crisp) {
            this.parent = crisp;
        }
        /**
         * Sets the user nickname
         */
        CrispUser.prototype.setNickname = function (nickname) {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["set", "user:nickname", [nickname]]);
        };
        /**
         * Sets the user email
         */
        CrispUser.prototype.setEmail = function (email, hmac) {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["set", "user:email", [email, hmac]]);
        };
        /**
         * Sets the user phone
         */
        CrispUser.prototype.setPhone = function (phone) {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["set", "user:phone", [phone]]);
        };
        /**
         * Sets the user avatar
         */
        CrispUser.prototype.setAvatar = function (avatar) {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["set", "user:avatar", [avatar]]);
        };
        /**
         * Sets the user company
         */
        CrispUser.prototype.setCompany = function (name, data) {
            var payload = {};
            if (data === null || data === void 0 ? void 0 : data.url) {
                payload.url = data.url;
            }
            if (data === null || data === void 0 ? void 0 : data.description) {
                payload.description = data.description;
            }
            if (data === null || data === void 0 ? void 0 : data.employment) {
                payload.employment = [
                    data.employment.title,
                    data.employment.role
                ];
            }
            if (data === null || data === void 0 ? void 0 : data.geolocation) {
                payload.geolocation = [data.geolocation.country];
                if (data.geolocation.city) {
                    payload.geolocation.push(data.geolocation.city);
                }
            }
            this.parent.createSingletonIfNecessary();
            if (Object.keys(payload).length > 0) {
                window.$crisp.push(["set", "user:company", [name, payload]]);
            }
            else {
                window.$crisp.push(["set", "user:company", [name]]);
            }
        };
        /**
         * Gets the user email
         */
        CrispUser.prototype.getEmail = function () {
            if (!this.parent.isCrispInjected()) {
                return null;
            }
            return window.$crisp.get("user:email");
        };
        /**
         * Gets the user phone
         */
        CrispUser.prototype.getPhone = function () {
            if (!this.parent.isCrispInjected()) {
                return null;
            }
            return window.$crisp.get("user:phone");
        };
        /**
         * Gets the user nickname
         */
        CrispUser.prototype.getNickname = function () {
            if (!this.parent.isCrispInjected()) {
                return null;
            }
            return window.$crisp.get("user:nickname");
        };
        /**
         * Gets the user avatar
         */
        CrispUser.prototype.getAvatar = function () {
            if (!this.parent.isCrispInjected()) {
                return null;
            }
            return window.$crisp.get("user:avatar");
        };
        /**
         * Gets the user company
         */
        CrispUser.prototype.getCompany = function () {
            if (!this.parent.isCrispInjected()) {
                return null;
            }
            return window.$crisp.get("user:company");
        };
        /**
         * Registers a callback for email changed event
         */
        CrispUser.prototype.onEmailChanged = function (callback) {
            if (this.parent.isCrispInjected()) {
                this.offEmailChanged();
                window.$crisp.push(["on", "user:email:changed", callback]);
            }
        };
        /**
         * Unregisters the email changed callback
         */
        CrispUser.prototype.offEmailChanged = function () {
            if (this.parent.isCrispInjected()) {
                window.$crisp.push(["off", "user:email:changed"]);
            }
        };
        /**
         * Registers a callback for phone changed event
         */
        CrispUser.prototype.onPhoneChanged = function (callback) {
            if (this.parent.isCrispInjected()) {
                this.offPhoneChanged();
                window.$crisp.push(["on", "user:phone:changed", callback]);
            }
        };
        /**
         * Unregisters the phone changed callback
         */
        CrispUser.prototype.offPhoneChanged = function () {
            if (this.parent.isCrispInjected()) {
                window.$crisp.push(["off", "user:phone:changed"]);
            }
        };
        /**
         * Registers a callback for nickname changed event
         */
        CrispUser.prototype.onNicknameChanged = function (callback) {
            if (this.parent.isCrispInjected()) {
                this.offNicknameChanged();
                window.$crisp.push(["on", "user:nickname:changed", callback]);
            }
        };
        /**
         * Unregisters the nickname changed callback
         */
        CrispUser.prototype.offNicknameChanged = function () {
            if (this.parent.isCrispInjected()) {
                window.$crisp.push(["off", "user:nickname:changed"]);
            }
        };
        /**
         * Registers a callback for avatar changed event
         */
        CrispUser.prototype.onAvatarChanged = function (callback) {
            if (this.parent.isCrispInjected()) {
                this.offAvatarChanged();
                window.$crisp.push(["on", "user:avatar:changed", callback]);
            }
        };
        /**
         * Unregisters the avatar changed callback
         */
        CrispUser.prototype.offAvatarChanged = function () {
            if (this.parent.isCrispInjected()) {
                window.$crisp.push(["off", "user:avatar:changed"]);
            }
        };
        return CrispUser;
    }());

    /*
     * This file is part of crisp-sdk-web
     *
     * Copyright (c) 2025 Crisp IM SAS
     * All rights belong to Crisp IM SAS
     */
    /**************************************************************************
     * CLASS
     ***************************************************************************/
    /**
     * Crisp trigger management
     */
    var CrispTrigger = /** @class */ (function () {
        /**
         * Constructor
         */
        function CrispTrigger(crisp) {
            this.parent = crisp;
        }
        /**
         * Runs a trigger by name
         */
        CrispTrigger.prototype.run = function (name) {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["do", "trigger:run", [name]]);
        };
        return CrispTrigger;
    }());

    /*
     * This file is part of crisp-sdk-web
     *
     * Copyright (c) 2025 Crisp IM SAS
     * All rights belong to Crisp IM SAS
     */
    /**
     * Crisp scenario management
     */
    var CrispScenario = /** @class */ (function () {
        /**
         * Constructor
         */
        function CrispScenario(crisp) {
            this.parent = crisp;
        }
        /**
         * Runs a bot scenario by name, with optional template variables
         */
        CrispScenario.prototype.run = function (name, variables) {
            this.parent.createSingletonIfNecessary();
            if (variables && typeof variables === "object") {
                window.$crisp.push(["do", "bot:scenario:run", [name, variables]]);
            }
            else {
                window.$crisp.push(["do", "bot:scenario:run", [name]]);
            }
        };
        return CrispScenario;
    }());

    /*
     * This file is part of crisp-sdk-web
     *
     * Copyright (c) 2025 Crisp IM SAS
     * All rights belong to Crisp IM SAS
     */
    /**************************************************************************
     * ENUMERATIONS
     ***************************************************************************/
    /* eslint-disable no-unused-vars */
    exports.EventsColors = void 0;
    (function (EventsColors) {
        EventsColors["Red"] = "red";
        EventsColors["Orange"] = "orange";
        EventsColors["Yellow"] = "yellow";
        EventsColors["Green"] = "green";
        EventsColors["Blue"] = "blue";
        EventsColors["Purple"] = "purple";
        EventsColors["Pink"] = "pink";
        EventsColors["Brown"] = "brown";
        EventsColors["Grey"] = "grey";
        EventsColors["Black"] = "black";
    })(exports.EventsColors || (exports.EventsColors = {}));
    /* eslint-enable no-unused-vars */
    /**************************************************************************
     * CLASS
     ***************************************************************************/
    /**
     * Crisp session management
     */
    var CrispSession = /** @class */ (function () {
        /**
         * Constructor
         */
        function CrispSession(crisp) {
            this.parent = crisp;
        }
        /**
         * Resets the current session
         */
        CrispSession.prototype.reset = function (reload) {
            if (reload === void 0) { reload = false; }
            if (this.parent.isCrispInjected()) {
                window.$crisp.push(["do", "session:reset", [reload]]);
            }
        };
        /**
         * Sets session segments
         */
        CrispSession.prototype.setSegments = function (segments, overwrite) {
            if (overwrite === void 0) { overwrite = false; }
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["set", "session:segments", [segments, overwrite]]);
        };
        /**
         * Sets session data
         */
        CrispSession.prototype.setData = function (data) {
            var _this = this;
            var payload = [];
            Object.entries(data).forEach(function (item) {
                if (_this.isValidDataValue(item[1])) {
                    payload.push([item[0], item[1]]);
                }
            });
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["set", "session:data", [payload]]);
        };
        /**
         * Pushes an event to the session
         */
        CrispSession.prototype.pushEvent = function (name, data, color) {
            if (data === void 0) { data = {}; }
            if (color === void 0) { color = exports.EventsColors.Blue; }
            if (this.parent.isCrispInjected()) {
                window.$crisp.push(["set", "session:event", [[[name, data, color]]]]);
            }
        };
        /**
         * Gets session data by key
         */
        CrispSession.prototype.getData = function (key) {
            if (!this.parent.isCrispInjected()) {
                return undefined;
            }
            return window.$crisp.get("session:data", key);
        };
        /**
         * Gets the session identifier
         */
        CrispSession.prototype.getIdentifier = function () {
            if (!this.parent.isCrispInjected()) {
                return null;
            }
            return window.$crisp.get("session:identifier");
        };
        /**
         * Registers a callback for session loaded event
         */
        CrispSession.prototype.onLoaded = function (callback) {
            this.parent.createSingletonIfNecessary();
            this.offLoaded();
            window.$crisp.push(["on", "session:loaded", callback]);
        };
        /**
         * Unregisters the session loaded callback
         */
        CrispSession.prototype.offLoaded = function () {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["off", "session:loaded"]);
        };
        /**
         * Checks if a value is valid for session data
         */
        CrispSession.prototype.isValidDataValue = function (value) {
            return (typeof value === "string" ||
                typeof value === "number" ||
                typeof value === "boolean");
        };
        return CrispSession;
    }());

    /*
     * This file is part of crisp-sdk-web
     *
     * Copyright (c) 2025 Crisp IM SAS
     * All rights belong to Crisp IM SAS
     */
    /**************************************************************************
     * CLASS
     ***************************************************************************/
    /**
     * Crisp chat management
     */
    var CrispChat = /** @class */ (function () {
        /**
         * Constructor
         */
        function CrispChat(crisp) {
            this.parent = crisp;
        }
        /**
         * Shows the chat widget
         */
        CrispChat.prototype.show = function () {
            this.parent.autoInjectIfNecessary();
            window.$crisp.push(["do", "chat:show"]);
        };
        /**
         * Hides the chat widget
         */
        CrispChat.prototype.hide = function () {
            this.parent.autoInjectIfNecessary();
            window.$crisp.push(["do", "chat:hide"]);
        };
        /**
         * Opens the chat widget
         */
        CrispChat.prototype.open = function () {
            this.parent.autoInjectIfNecessary();
            window.$crisp.push(["do", "chat:open"]);
        };
        /**
         * Closes the chat widget
         */
        CrispChat.prototype.close = function () {
            if (this.parent.isCrispInjected()) {
                window.$crisp.push(["do", "chat:close"]);
            }
        };
        /**
         * Sets an overlay view with a custom callback
         */
        CrispChat.prototype.setOverlayView = function (callback) {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["do", "overlay:open", [callback]]);
        };
        /**
         * Sets the helpdesk view
         */
        CrispChat.prototype.setHelpdeskView = function () {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["do", "helpdesk:search"]);
        };
        /**
         * Opens a helpdesk article
         */
        CrispChat.prototype.openHelpdeskArticle = function (locale, slug, title, category) {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["do", "helpdesk:article:open", [
                    locale,
                    slug,
                    title,
                    category
                ]]);
        };
        /**
         * Queries the helpdesk
         */
        CrispChat.prototype.queryHelpdesk = function (query) {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["do", "helpdesk:query", [query]]);
        };
        /**
         * Gets the unread message count
         */
        CrispChat.prototype.unreadCount = function () {
            if (!this.parent.isCrispInjected()) {
                return 0;
            }
            return window.$crisp.get("chat:unread:count");
        };
        /**
         * Checks if the chat is opened
         */
        CrispChat.prototype.isChatOpened = function () {
            if (!this.parent.isCrispInjected()) {
                return false;
            }
            return window.$crisp.is("chat:opened");
        };
        /**
         * Checks if the chat is visible
         */
        CrispChat.prototype.isVisible = function () {
            if (!this.parent.isCrispInjected()) {
                return false;
            }
            return window.$crisp.is("chat:visible");
        };
        /**
         * Registers a callback for chat initiated event
         */
        CrispChat.prototype.onChatInitiated = function (callback) {
            this.parent.createSingletonIfNecessary();
            this.offChatInitiated();
            window.$crisp.push(["on", "chat:initiated", callback]);
        };
        /**
         * Unregisters the chat initiated callback
         */
        CrispChat.prototype.offChatInitiated = function () {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["off", "chat:initiated"]);
        };
        /**
         * Registers a callback for chat opened event
         */
        CrispChat.prototype.onChatOpened = function (callback) {
            this.parent.createSingletonIfNecessary();
            this.offChatOpened();
            window.$crisp.push(["on", "chat:opened", callback]);
        };
        /**
         * Unregisters the chat opened callback
         */
        CrispChat.prototype.offChatOpened = function () {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["off", "chat:opened"]);
        };
        /**
         * Registers a callback for chat closed event
         */
        CrispChat.prototype.onChatClosed = function (callback) {
            this.parent.createSingletonIfNecessary();
            this.offChatClosed();
            window.$crisp.push(["on", "chat:closed", callback]);
        };
        /**
         * Unregisters the chat closed callback
         */
        CrispChat.prototype.offChatClosed = function () {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["off", "chat:closed"]);
        };
        /**
         * Registers a callback for helpdesk queried event
         */
        CrispChat.prototype.onHelpdeskQueried = function (callback) {
            this.parent.createSingletonIfNecessary();
            this.offHelpdeskQueried();
            window.$crisp.push(["on", "helpdesk:queried", callback]);
        };
        /**
         * Unregisters the helpdesk queried callback
         */
        CrispChat.prototype.offHelpdeskQueried = function () {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["off", "helpdesk:queried"]);
        };
        return CrispChat;
    }());

    /*
     * This file is part of crisp-sdk-web
     *
     * Copyright (c) 2025 Crisp IM SAS
     * All rights belong to Crisp IM SAS
     */
    /* eslint-enable no-unused-vars */
    /**************************************************************************
     * CONSTANTS
     ***************************************************************************/
    var TOOL_EVENT_PREFIX = "hugo:tool:";
    /**************************************************************************
     * CLASS
     ***************************************************************************/
    /**
     * Crisp Hugo (AI agent) management
     */
    var CrispHugo = /** @class */ (function () {
        /**
         * Constructor
         */
        function CrispHugo(crisp) {
            this.parent = crisp;
        }
        /**
         * Registers a handler for a Hugo Widget Tool
         */
        CrispHugo.prototype.onTool = function (name, handler) {
            this.parent.createSingletonIfNecessary();
            this.offTool(name);
            window.$crisp.push(["on", "".concat(TOOL_EVENT_PREFIX).concat(name), handler]);
        };
        /**
         * Unregisters a previously registered Hugo Widget Tool handler
         */
        CrispHugo.prototype.offTool = function (name) {
            this.parent.createSingletonIfNecessary();
            window.$crisp.push(["off", "".concat(TOOL_EVENT_PREFIX).concat(name)]);
        };
        return CrispHugo;
    }());

    /*
     * This file is part of crisp-sdk-web
     *
     * Copyright (c) 2025 Crisp IM SAS
     * All rights belong to Crisp IM SAS
     */
    /**************************************************************************
     * IMPORTS
     ***************************************************************************/
    // PROJECT: MAIN
    /**************************************************************************
     * ENUMERATIONS
     ***************************************************************************/
    /* eslint-disable no-unused-vars */
    exports.ChatboxColors = void 0;
    (function (ChatboxColors) {
        ChatboxColors["Default"] = "default";
        ChatboxColors["Amber"] = "amber";
        ChatboxColors["Black"] = "black";
        ChatboxColors["Blue"] = "blue";
        ChatboxColors["BlueGrey"] = "blue_grey";
        ChatboxColors["LightBlue"] = "light_blue";
        ChatboxColors["Brown"] = "brown";
        ChatboxColors["Cyan"] = "cyan";
        ChatboxColors["Green"] = "green";
        ChatboxColors["LightGreen"] = "light_green";
        ChatboxColors["Grey"] = "grey";
        ChatboxColors["Indigo"] = "indigo";
        ChatboxColors["Orange"] = "orange";
        ChatboxColors["DeepOrange"] = "deep_orange";
        ChatboxColors["Pink"] = "pink";
        ChatboxColors["Purple"] = "purple";
        ChatboxColors["DeepPurple"] = "deep_purple";
        ChatboxColors["Red"] = "red";
        ChatboxColors["Teal"] = "teal";
    })(exports.ChatboxColors || (exports.ChatboxColors = {}));
    exports.ChatboxPosition = void 0;
    (function (ChatboxPosition) {
        ChatboxPosition["Left"] = "left";
        ChatboxPosition["Right"] = "right";
    })(exports.ChatboxPosition || (exports.ChatboxPosition = {}));
    /* eslint-enable no-unused-vars */
    /**************************************************************************
     * CLASS
     ***************************************************************************/
    /**
     * Main Crisp SDK class
     */
    var Crisp = /** @class */ (function () {
        /**
         * Constructor
         */
        function Crisp() {
            this.clientUrl = "https://client.crisp.chat/l.js";
            this.websiteId = "";
            this.autoload = true;
            this.injected = false;
            this.chat = new CrispChat(this);
            this.session = new CrispSession(this);
            this.user = new CrispUser(this);
            this.message = new CrispMessage(this);
            this.trigger = new CrispTrigger(this);
            this.scenario = new CrispScenario(this);
            this.hugo = new CrispHugo(this);
        }
        /**
         * Configures the Crisp SDK
         */
        Crisp.prototype.configure = function (websiteId, options) {
            if (options === void 0) { options = {}; }
            this.websiteId = websiteId;
            this.tokenId = options.tokenId;
            this.locale = options.locale;
            this.sessionMerge = options.sessionMerge;
            this.cookieDomain = options.cookieDomain;
            this.cookieExpire = options.cookieExpire;
            this.lockFullview = options.lockFullview;
            this.lockMaximized = options.lockMaximized;
            this.safeMode = options.safeMode;
            if (options.clientUrl !== undefined) {
                this.clientUrl = options.clientUrl;
            }
            if (options.autoload !== undefined) {
                this.autoload = options.autoload;
            }
            if (this.autoload) {
                this.load();
            }
        };
        /**
         * Loads the Crisp client script
         */
        Crisp.prototype.load = function () {
            var head = document.getElementsByTagName("head");
            this.createSingletonIfNecessary();
            if (this.isCrispInjected() === true) {
                return;
            }
            if (!this.websiteId) {
                throw new Error("websiteId must be set before loading Crisp");
            }
            window.CRISP_WEBSITE_ID = this.websiteId;
            window.CRISP_RUNTIME_CONFIG = {};
            if (this.tokenId) {
                window.CRISP_TOKEN_ID = this.tokenId;
            }
            if (this.sessionMerge) {
                window.CRISP_RUNTIME_CONFIG.session_merge = true;
            }
            if (this.locale) {
                window.CRISP_RUNTIME_CONFIG.locale = this.locale;
            }
            if (this.lockFullview) {
                window.CRISP_RUNTIME_CONFIG.lock_full_view = true;
            }
            if (this.lockMaximized) {
                window.CRISP_RUNTIME_CONFIG.lock_maximized = true;
            }
            if (this.cookieDomain) {
                window.CRISP_COOKIE_DOMAIN = this.cookieDomain;
            }
            if (this.cookieExpire) {
                window.CRISP_COOKIE_EXPIRE = this.cookieExpire;
            }
            if (!head || !head[0]) {
                return this.deferredLoading();
            }
            if (this.safeMode === true) {
                this.setSafeMode(true);
            }
            var script = document.createElement("script");
            script.src = this.clientUrl;
            script.async = true;
            head[0].appendChild(script);
            this.injected = true;
        };
        /**
         * Sets the token ID for authentication
         */
        Crisp.prototype.setTokenId = function (tokenId) {
            this.tokenId = tokenId;
            if (this.isCrispInjected() === true) {
                if (tokenId) {
                    window.CRISP_TOKEN_ID = tokenId;
                }
                else {
                    delete window.CRISP_TOKEN_ID;
                }
            }
        };
        /**
         * Sets the z-index of the chat widget
         */
        Crisp.prototype.setZIndex = function (zIndex) {
            this.createSingletonIfNecessary();
            window.$crisp.push(["config", "container:index", [zIndex]]);
        };
        /**
         * Sets the color theme of the chat widget
         */
        Crisp.prototype.setColorTheme = function (color) {
            this.createSingletonIfNecessary();
            window.$crisp.push(["config", "color:theme", [color]]);
        };
        /**
         * Sets whether to hide the widget when operators are away
         */
        Crisp.prototype.setHideOnAway = function (enabled) {
            this.createSingletonIfNecessary();
            window.$crisp.push(["config", "hide:on:away", [enabled]]);
        };
        /**
         * Sets whether to hide the widget on mobile devices
         */
        Crisp.prototype.setHideOnMobile = function (enabled) {
            this.createSingletonIfNecessary();
            window.$crisp.push(["config", "hide:on:mobile", [enabled]]);
        };
        /**
         * Sets the position of the chat widget
         */
        Crisp.prototype.setPosition = function (position) {
            this.createSingletonIfNecessary();
            window.$crisp.push(["config", "position:reverse", [
                    position === exports.ChatboxPosition.Left
                ]]);
        };
        /**
         * Sets whether to show the availability tooltip
         */
        Crisp.prototype.setAvailabilityTooltip = function (enabled) {
            this.createSingletonIfNecessary();
            window.$crisp.push(["config", "availability:tooltip", [enabled]]);
        };
        /**
         * Sets whether vacation mode is enabled
         */
        Crisp.prototype.setVacationMode = function (enabled) {
            this.createSingletonIfNecessary();
            window.$crisp.push(["config", "hide:vacation", [enabled]]);
        };
        /**
         * Sets safe mode for error handling
         */
        Crisp.prototype.setSafeMode = function (safe) {
            if (safe === void 0) { safe = true; }
            this.createSingletonIfNecessary();
            window.$crisp.push(["safe", safe]);
        };
        /**
         * Mutes or unmutes notification sounds
         */
        Crisp.prototype.muteSound = function (mute) {
            this.createSingletonIfNecessary();
            window.$crisp.push(["config", "sound:mute", [mute]]);
        };
        /**
         * Toggles the operator count display
         */
        Crisp.prototype.toggleOperatorCount = function (enabled) {
            this.createSingletonIfNecessary();
            window.$crisp.push(["config", "show:operator:count", [enabled]]);
        };
        /**
         * Registers a callback for website availability changed event
         */
        Crisp.prototype.onWebsiteAvailabilityChanged = function (callback) {
            this.createSingletonIfNecessary();
            this.offWebsiteAvailabilityChanged();
            window.$crisp.push(["on", "website:availability:changed", callback]);
        };
        /**
         * Unregisters the website availability changed callback
         */
        Crisp.prototype.offWebsiteAvailabilityChanged = function () {
            this.createSingletonIfNecessary();
            window.$crisp.push(["off", "website:availability:changed"]);
        };
        /**
         * Creates the $crisp singleton if it does not exist
         */
        Crisp.prototype.createSingletonIfNecessary = function () {
            if (window.$crisp === undefined) {
                window.$crisp = [];
            }
        };
        /**
         * Auto-injects the Crisp client if not already injected
         */
        Crisp.prototype.autoInjectIfNecessary = function () {
            if (!this.isCrispInjected()) {
                this.load();
            }
        };
        /**
         * Checks if the Crisp client is injected
         */
        Crisp.prototype.isCrispInjected = function () {
            var _a;
            if (this.injected === true || ((_a = window.$crisp) === null || _a === void 0 ? void 0 : _a.is)) {
                return true;
            }
            return false;
        };
        /**
         * Defers loading until DOM is ready
         */
        Crisp.prototype.deferredLoading = function () {
            var _this = this;
            document.addEventListener("DOMContentLoaded", function () {
                _this.load();
            });
        };
        return Crisp;
    }());
    /**************************************************************************
     * EXPORTS
     ***************************************************************************/
    var singleton = new Crisp();

    exports.Crisp = singleton;
    exports.CrispClass = Crisp;

}));
